src/raftstore/store/fsm/apply.rs

```diff
@@ -2029,196 +2084,474 @@ pub struct ApplyRes {
 
-        if !core.apply_res.is_empty() {
-            self.notifier
-                .send(TaskRes::Applies(core.apply_res))
-                .unwrap();
+        if ctx.timer.is_none() {
+            ctx.timer = Some(SlowTimer::new());
+        }
+        if !state.pending_entries.is_empty() {
+            self.delegate
+                .handle_raft_committed_entries(ctx, state.pending_entries);
+            if let Some(ref mut s) = self.delegate.wait_merge_state {
+                // So the delegate is executing another `CommitMerge`.
+                s.pending_msgs = state.pending_msgs;
+                s.catch_up_logs = state.catch_up_logs;
+                return false;
+            }
         }
 
-        STORE_APPLY_LOG_HISTOGRAM.observe(duration_to_sec(t.elapsed()) as f64);
+        if !state.pending_msgs.is_empty() {
+            self.handle_tasks(ctx, &mut state.pending_msgs);
+        }
 
-        slow_log!(
-            t,
-            "{} handle ready {} committed entries",
-            self.tag,
-            core.committed_count
-        );
+        match self.delegate.wait_merge_state {
+            // So the delegate is executing another `CommitMerge` when handling
+            // `pending_msgs`.
+            Some(ref mut s) => {
+                s.catch_up_logs = state.catch_up_logs;
+                false
+            }
+            None => {
+                info!("{} all pending logs are applied.", self.delegate.tag);
+                if let Some(mut catch_up_logs) = state.catch_up_logs {
+                    ctx.write_to_db();
+                    let region_id = self.delegate.region_id();
+                    catch_up_logs
+                        .ready_to_merge
+                        .store(region_id, Ordering::SeqCst);
+                    let _ = catch_up_logs
+                        .target_mailbox
+                        .force_send(Msg::LogsUpToDate(region_id));
+                }
+                true
+            }
+        }
     }
 
-    fn handle_proposals(&mut self, proposals: Vec<RegionProposal>) {
-        let mut propose_num = 0;
-        for region_proposal in proposals {
-            propose_num += region_proposal.props.len();
-            let delegate = match self.delegates.get_mut(&region_proposal.region_id) {
-                Some(d) => d.as_mut().unwrap(),
+    fn catch_up_logs_for_merge(&mut self, ctx: &mut ApplyContext, mut catch_up_logs: CatchUpLogs) {
+        if ctx.timer.is_none() {
+            ctx.timer = Some(SlowTimer::new());
+        }
+        let apply_index = self.delegate.apply_state.get_applied_index();
+        'check: {
+            if apply_index >= catch_up_logs.merge.get_commit() {
+                break 'check;
+            }
+
+            let entries =
+                self.delegate
+                    .load_entries_for_merge(ctx, &catch_up_logs.merge, apply_index);
+            if entries.is_empty() {
+                panic!(
```

Comment:
we don't check if the source peer up to date when committing merge now, so this function will always be called for merge process, right? What if the source peer is already up to date.

Reply:
If it's update to date, it will break at L2271.